Networking/Networking/Remote/MediaRemote.swift

```diff
@@ -94,6 +117,35 @@ public class MediaRemote: Remote, MediaRemoteProtocol {
         enqueue(request, mapper: mapper, completion: completion)
     }
 
+    /// Loads an array of media from the site's WP Media Library via WordPress site API.
+    /// API reference: https://developer.wordpress.org/rest-api/reference/media/#list-media
+    ///
+    /// - Parameters:
+    ///   - siteURL: Site for which we'll load the media from.
+    ///   - pageNumber: The index of the page of media data to load from, starting from 1.
+    ///   - pageSize: The number of media items to return.
+    ///   - completion: Closure to be executed upon completion.
+    public func loadMediaLibraryUsingRestApi(siteURL: String,
+                                             pageNumber: Int = Default.pageNumber,
+                                             pageSize: Int = 25,
+                                             completion: @escaping (Result<[WordPressMedia], Error>) -> Void) {
```

Comment:
We already have two ways of making media requests to WPCOM and directly to the site. Is there another way to make the REST request work with the existing `loadMediaLibraryFromWordPressSite` and `uploadMediaToWordPressSite`? All the code looks the same until `request = RESTRequest/DotcomRequest`. The difference between the two is authentication, an implementation detail in the networking layer, it'd be great if we can support REST requests without changing the protocol in the networking layer. I'm not very familiar with the implementation, maybe some way to make a `DotcomRequest` convertible to `RESTRequest` would work like `JetpackRequest.asRESTRequest(with:)`.

Reply:
> The difference between the two is authentication, an implementation detail in the networking layer

- They have different paths. [DotCom Path](https://github.com/woocommerce/woocommerce-ios/blob/ff79bf90d87a3fe00849d98cf43d79692333534e/Networking/Networking/Remote/MediaRemote.swift#L110) vs [REST path](https://github.com/woocommerce/woocommerce-ios/blob/ff79bf90d87a3fe00849d98cf43d79692333534e/Networking/Networking/Remote/MediaRemote.swift#L139)
- One uses [`siteID`](https://github.com/woocommerce/woocommerce-ios/blob/ff79bf90d87a3fe00849d98cf43d79692333534e/Networking/Networking/Remote/MediaRemote.swift#L99) and the other uses [`siteURL`](https://github.com/woocommerce/woocommerce-ios/blob/ff79bf90d87a3fe00849d98cf43d79692333534e/Networking/Networking/Remote/MediaRemote.swift#L128)

--- 

> maybe some way to make a DotcomRequest convertible to RESTRequest would work like JetpackRequest.asRESTRequest(with:).

I initially thought about converting `DotcomRequest` into `RESTRequest`. [Old deleted PR](https://github.com/woocommerce/woocommerce-ios/pull/8572)

But, when I discussed with Hicham p1673011537776019/1672921523.522859-slack-C046HDZL87J I realised that it would be confusing to convert `DotcomRequest` into `RESTRequest`

In the approach I am following right now, we will add another condition in `MediaStore` to decide between connecting using siteID or siteURL. (Just like the JCP check)

Even though we have to add new methods in all layers, this approach makes more sense.
-  Converting `DotcomRequest` into `RESTRequest` isn't a common requirement. As far as I know, the Media remote would be the only place we need it. 
- Converting `DotcomRequest` into `RESTRequest` isn't straightforward. The path isn't the same between `DotcomRequest` and `RESTRequest`. We will have to manually remove the `/sites` path like [this](https://github.com/woocommerce/woocommerce-ios/pull/8572/files#diff-919cee6a4f7d4e5c7ab0dbb24c5ecca1ee151a11dff3c56ac2eed8f23623b50dR99-R103). 

--- 

If you think seeing the whole implementation will help, let us keep this PR open until I raise all the child PRs. 🙏

Reply:
@jaclync To show you the whole picture, I have created the following child PRs

- For choosing between site URL and site ID in `MediaStore` - https://github.com/woocommerce/woocommerce-ios/pull/8603
-  Migrated `ProductsRemote` (Not directly related to our discussion) - https://github.com/woocommerce/woocommerce-ios/pull/8605

Reply:
Thanks for the links, sorry I might not be getting the full picture after glancing through the Slack discussion 😅 My major concern is that the REST implementation details now leak to the view layer, like how the [view data source now needs to know about the default credentials](https://github.com/woocommerce/woocommerce-ios/pull/8603/files#diff-4ff03b3a1050100b2fe7b6ed6866bb083d85db08cae3416c4df9a6cc63494ec5R165-R171). For direct media requests to the site for JCP sites, [the check is within the Yosemite layer](https://github.com/woocommerce/woocommerce-ios/blob/41756a5e4eab9e091697814bc43ccd9e9aa19b6f/Yosemite/Yosemite/Stores/MediaStore.swift#L73-L80). I was confused about why we can do all the REST request URL conversion & authentication using `RequestConverter` and `RequestAuthenticator` within the networking layer for all other Jetpack -> REST requests, but not for this use case.

If a separate remote call is still necessary, I'd suggest sharing the same code for the request parameters by moving it into a function if possible.

@itsmeichigo please feel free to take over the code review, there are probably some context and details that I'm missing and I don't want to block the progress.

Reply:
> I was confused about why we can do all the REST request URL conversion & authentication using RequestConverter and RequestAuthenticator within the networking layer for all other Jetpack -> REST requests, but not for this use case.

It is possible to convert `DotcomRequest` into `RESTRequest` as I tried in [this deleted PR](https://github.com/woocommerce/woocommerce-ios/pull/8572)

I will explain my view below about why I think converting  `DotcomRequest` into `RESTRequest` in the network layer isn't a good solution. 

- `DotcomRequest`s are processed by WPCOM servers, whereas all `JetpackRequest`s are forwarded to the site by WPCOM servers (using Jetpack tunnel). 
- If we create a way to convert `DotcomRequest`s into `JetpackRequest`s that would create an illusion that any `DotcomRequest` can be converted into REST request, authenticated using application password and forwarded to the site. But this isn't true. `DotcomRequest`s are meant to be solely processed by WPCOM servers.

----

> My major concern is that the REST implementation details now leak to the view layer

Even if we auto convert `DotcomRequest` into `RESTRequest`, we still have to pass a value from the view layer into Yosemite to force the `MediaStore` to always choose `WordPressSite` method. i.e. Pass a value like `forceRestApi` along with `MediaAction` upon which the `MediaStore` can skip the JCP checks and choose `WordPressSite` methods by default.

This means that we have to leak this REST implementation into the view layer in both cases. 

---

Thanks for your input, Jaclyn! 🙇 Please correct me if I am wrong.

Reply:
Thanks, Sharma for the detailed explanations! My apologies for not actively participating in that Slack thread, but I want to share my opinion on this:

> If we create a way to convert DotcomRequests into JetpackRequests that would create an illusion that any DotcomRequest can be converted into REST request,

If we follow what we are doing with `JetpackRequest` by adding `availableAsRESTRequest`, I think it would be pretty clear to understand that not all `DotcomRequest` can be converted to REST request. Plus, it would be so much simpler if we follow the same route as what we did with `JetpackRequest` because we can get the site address from the credentials in `AlamofireNetwork`. 

Do you think it's worth it to switch to that approach instead?

Reply:
Thank you for your input, @itsmeichigo! 🙏 

> If we follow what we are doing with JetpackRequest by adding availableAsRESTRequest, I think it would be pretty clear to understand that not all DotcomRequest can be converted into a REST request.

IMHO, by adding `availableAsRESTRequest` parameter and making `DotcomRequest` convertible into `RESTRequest` would open up possibilities for future confusion and potential bugs. I would prefer to avoid it mainly because we add that parameter just for this single usecase (`MediaRemote`)

On the other hand, I understand that by adding `availableAsRESTRequest` parameter to `DotcomRequest` will reduce the number of code changes needed in `MediaStore` and `MediaRemote`. (BTW, we will still have to inject `forceRestApi: Bool` from the  view layer if logged in using `wporg` creds)

I am good with switching to that approach if we as a team prefer doing that. Actually, I have [this deleted PR](https://github.com/woocommerce/woocommerce-ios/pull/8572) in which I tried that approach. I am not happy with doing [the trimming of path](https://github.com/woocommerce/woocommerce-ios/pull/8572/files#diff-919cee6a4f7d4e5c7ab0dbb24c5ecca1ee151a11dff3c56ac2eed8f23623b50dR72-R91), whereas for a `JetpackRequest` [creating a `RESTRequest` is straightforward](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Networking/Networking/Requests/JetpackRequest.swift#L85-L91). 

---

I will close this PR work using the "convert `DotcomRequest`" approach next week.  
Thank you, @jaclync and @itsmeichigo, for the review! 🙇

Reply:
> BTW, we will still have to inject forceRestApi: Bool from the view layer if logged in using wporg creds

@selanthiraiyan Is it possible to check for `siteID` - if it's a constant site ID we can force using REST API and skip JCP check? Does that sound too hacky though?

Reply:
> Even if we auto convert DotcomRequest into RESTRequest, we still have to pass a value from the view layer into Yosemite to force the MediaStore to always choose WordPressSite method. i.e. Pass a value like forceRestApi along with MediaAction upon which the MediaStore can skip the JCP checks and choose WordPressSite methods by default.
> This means that we have to leak this REST implementation into the view layer in both cases.

Can we avoid leaking the REST implementation to the view layer by checking the type of credentials at the Yosemite/Networking layer instead of passing a `forceRestApi` flag from the view layer?

Reply:
> Is it possible to check for siteID - if it's a constant site ID we can force using REST API and skip JCP check? Does that sound too hacky though?

@itsmeichigo 
Good idea. I think we can stick to this plan for now. We can work on a different solution if we implement store-switching capabilities. 

> Can we avoid leaking the REST implementation to the view layer by checking the type of credentials at the Yosemite/Networking layer instead of passing a forceRestApi flag from the view layer?

@jaclync The credentials aren't available inside the `Yosemite` layer (`MediaStore`) from which we have to check the condition and choose between endpoints.